types/coin.go

```diff
@@ -240,51 +240,34 @@ func (coins Coins) Add(coinsB ...Coin) Coins {
 // denomination and addition only occurs when the denominations match, otherwise
 // the coin is simply added to the sum assuming it's not zero.
 func (coins Coins) safeAdd(coinsB Coins) Coins {
-	sum := ([]Coin)(nil)
-	indexA, indexB := 0, 0
-	lenA, lenB := len(coins), len(coinsB)
-
-	for {
-		if indexA == lenA {
-			if indexB == lenB {
-				// return nil coins if both sets are empty
-				return sum
-			}
-
-			// return set B (excluding zero coins) if set A is empty
-			return append(sum, removeZeroCoins(coinsB[indexB:])...)
-		} else if indexB == lenB {
-			// return set A (excluding zero coins) if set B is empty
-			return append(sum, removeZeroCoins(coins[indexA:])...)
+	uniqSumByDenom := make(map[string]Coin)
+	combo := append(coins, coinsB...)
+	for _, coin := range combo {
```

Comment:
I would advise not to create `combo` and just iterate over both sets.

Reply:
That's going to create the need for
```go
for _, coinL := range []Coins{coins, coinsB} {
      for _, coin := range coinL {
             ...
      }
}
```
vs
```go
combo := append(coins, coinsB)
for _, coin := range combo {
     ...
}
```

which is more logic than needed IMHO for the purpose of combining both, which takes a readability and complexity hit.

Reply:
No, that's not what I meant. I meant iterate over the first set and then the second set.

Reply:
```go
for _, coin := range coins {
     // De-dupe logic here
}

for _, coin := range coinsB {
     // De-dupe logic here
}
```
this option didn't make sense to me which I why I presumed you meant `[]Coins{` given that now we have to dedupe twice, and can't assume that each Coins list has unique denominations. Did you mean something else though? Please perhaps help with pseudo code of what you meant.

Reply:
No that's exactly what I meant.

> can't assume that each Coins list has unique denominations

Actually, I believe you can and should make this assumption.

The main reason I want to do this is that I want to avoid the `append` which can grow capacity, copy and modify the underlying array. Just loop over each slice individually.

Reply:
> The main reason I want to do this is that I want to avoid the append which can grow capacity, copy and modify the underlying array. Just loop over each slice individually.

While that reasoning works for when modifying the content of combo by access the head of it, combo is only used to iterate and there isn't any modification going on. append here is guaranteed to create a new slice consisting of the union. I am hesitant to make that code duplication as I don't think it is necessary :)

> Actually, I believe you can and should make this assumption.

What happens for Coins sent in from the wild? The newer change is more code, more risk.

Reply:
It's a private auxiliary method -- validity happens upstream.

Reply:
> Actually, I believe you can and should make this assumption.

That's what we've done so far. It's on the client to ensure `Coins` validation actually takes place.

Reply:
This PR has been stalled for a while, am not comfortable going with that assumption of validation; @alessio and @alexanderbez would you like to open up a PR and implement that code? Feel free to adapt from this one, but the issue will need closing anyways, thanks.